packages/flutter/lib/src/widgets/selection.dart

```diff
@@ -0,0 +1,962 @@
+          } else if (newIndex == 0) {
+            finalResult = currentSelectableResult;
+          } else {
+            forward = false;
+            newIndex -= 1;
+          }
+          break;
+      }
+    }
+    if (isEnd) {
+      currentSelectionEndIndex = newIndex;
+    } else {
+      currentSelectionStartIndex = newIndex;
+    }
+    return finalResult!;
+  }
+}
+
+class _RenderSelectableContainerWidget extends RenderProxyBox with ChangeNotifier, Selectable, SelectionRegisteree {
+  _RenderSelectableContainerWidget({
+    required SelectionContainerDelegate? delegate,
+    required this.containerHasSelectableContent,
+    required SelectionRegistrar? selectionRegistrar,
+    RenderBox? child,
+  }) : _delegate = delegate,
+       super(child) {
+    _delegate?.addListener(notifyListeners);
+    containerHasSelectableContent.addListener(_handleContainerHasSelectableContentChange);
+    _handleContainerHasSelectableContentChange();
+    registrar = selectionRegistrar;
+  }
+
+  SelectionContainerDelegate? get delegate => _delegate;
+  SelectionContainerDelegate? _delegate;
+  set delegate(SelectionContainerDelegate? value) {
+    if (_delegate == null)
+      return;
+    _delegate?.removeListener(notifyListeners);
+    _delegate = value;
+    _delegate?.addListener(notifyListeners);
+  }
+
+  final ValueListenable<bool> containerHasSelectableContent;
+
+  void _handleContainerHasSelectableContentChange() {
+    hasSelectableContent.value = containerHasSelectableContent.value;
+  }
+
+  @override
+  void dispose() {
+    containerHasSelectableContent.removeListener(_handleContainerHasSelectableContentChange);
+    super.dispose();
+  }
+
+  @override
+  void pushHandleLayers(LayerLink? startHandle, LayerLink? endHandle) {
+    _delegate!.pushHandleLayers(startHandle, endHandle);
+  }
+
+  @override
+  Object? getSelectedContent() => _delegate!.getSelectedContent();
+
+  @override
+  SelectionResult dispatchSelectionEvent(SelectionEvent event) {
+    return _delegate!.dispatchSelectionEvent(event);
+  }
+
+  @override
+  SelectionGeometry? get value => _delegate!.value;
+}
+
+class _SelectionRegistrarScope extends InheritedWidget {
+  const _SelectionRegistrarScope({
+    Key? key,
+    required this.registrar,
+    required this.enabled,
+    required Widget child,
+  }) : super(key: key, child: child);
+
+  /// The registrar.
+  final SelectionRegistrar registrar;
+
+  /// whether this enabled
+  final bool enabled;
+
+  @override
+  bool updateShouldNotify(_SelectionRegistrarScope oldWidget) {
+    return oldWidget.registrar != registrar ||
+           oldWidget.enabled != enabled;
+  }
+}
+
+/// The adapter to handles the selections from the [SelectionArea] for
+/// [RichText].
+///
+/// By Wrapping a [RichText] with this widget, the [RichText] will be
+/// selectable under a [SelectionArea].
+///
+/// The [child] must be a [RichText].
+class TypographySelectionAdapter extends SingleChildRenderObjectWidget {
```

Comment:
@justinmc I am curious how you feel about this approach

To refactor the EditableText, the renderEditable needs to implement `RenderSelectableTypography` and EditableText needs to wrap the widget tree with this adapter.

These API can apply to non text renderobject too as long as they implement the `RenderSelectableTypography`, although a lot of class it uses has the keyword `Text*` in it, e.g TextSelection, TextRange, TextPosition, ideally these should have a more general names.

Reply:
`RenderSelectableTypography` is defined in rendering/selection.dart

Reply:
I am not entirely sure how to correctly handle RenderParagraph with WidgetSpan. The easiest we can do is to either select the entire widgetspan as one character, or make it possible to select partial content in the WidgetSpan. I know our internal customers like to inline text within text using WidgetSpan, and the former behavior may not suffice their needs.

Reply:
That seems super straightforward and clean to do, but I have some questions.

Is the ultimate point of having RenderEditable implement RenderSelectableTypography to allow the reuse of selection code?  I'll look more closely at what that would look like if so.

Also what's the status of TextFields mixing with SelectionAreas now?  I left repro steps in my main review about a bug related to this, but maybe it's not fully supported yet or something.

Reply:
If we want to support partial selection in WidgetSpan, this interface will not be useful because the Text widget will need to be some kind of SelectionContainer or the RenderParagraph needs to delegate the selection event down to individual widgetspan, but I think we don't need to support that in first version?

>Is the ultimate point of having RenderEditable implement RenderSelectableTypography to allow the reuse of selection code? I'll look more closely at what that would look like if so.

That is the current plan yes, but I started to worry about rich text editing, are we going to support something like widgetspan? if so, are we going to treat them as one character? can they be partially selected or modify? If so this interface won't work.

> Also what's the status of TextFields mixing with SelectionAreas now? I left repro steps in my main review about a bug related to this, but maybe it's not fully supported yet or something.

Right now they are two separate systems, so you can't select across text field with other text item. I think we should keep them separate even after refactoring. The selection can't go in or out from the textfield to other widget, otherwise text editing may be confusing. WDYT?

Reply:
Otherwise, IMO this API is perfect for renderobject without child and want to be selectable, it is far easier to implement this than the Selectable abstract class.